microbench/src/main/java/io/netty/microbench/search/SearchBenchmark.java

```diff
@@ -0,0 +1,214 @@
+            }
+        },
+        PREDICTABLE {
+            @Override
+            byte[] getNeedle(Random rnd) {
+                // all 0s
+                return new byte[64];
+            }
+            @Override
+            byte[] getHaystack(Random rnd) {
+                // no 0s except in the very end
+                byte[] bytes = randomBytes(rnd, 2048, 1, 255);
+                Arrays.fill(bytes, bytes.length - 64, bytes.length, (byte) 0);
+                return bytes;
+            }
+        },
+        UNPREDICTABLE {
+            @Override
+            byte[] getNeedle(Random rnd) {
+                return randomBytes(rnd, 64, 0, 1);
+            }
+            @Override
+            byte[] getHaystack(Random rnd) {
+                return randomBytes(rnd, 2048, 0, 1);
+            }
+        },
+        WORST_CASE { // ShiftingBitMask will fail on it because the needle is >64 bytes long
+            @Override
+            byte[] getNeedle(Random rnd) {
+                // aa(...)aab
+                byte[] needle = new byte[1024];
+                Arrays.fill(needle, (byte) 'a');
+                needle[needle.length - 1] = 'b';
+                return needle;
+            }
+            @Override
+            byte[] getHaystack(Random rnd) {
+                // aa(...)aaa
+                byte[] haystack = new byte[2048];
+                Arrays.fill(haystack, (byte) 'a');
+                return haystack;
+            }
+        };
+
+        abstract byte[] getNeedle(Random rnd);
+        abstract byte[] getHaystack(Random rnd);
+    }
+
+    @Param
+    public Input input;
+
+    @Param
+    public ByteBufType bufferType;
+
+    private Random rnd;
+    private ByteBuf needle, haystack;
+    private byte[] needleBytes, haystackBytes;
+    private SearchProcessorFactory kmpFactory, shiftingBitMaskFactory, ahoCorasicFactory;
+
+    @Setup
+    public void setup() {
+        rnd = new Random(SEED);
+
+        needleBytes = input.getNeedle(rnd);
+        haystackBytes = input.getHaystack(rnd);
+
+        needle = Unpooled.wrappedBuffer(needleBytes);
+        haystack = bufferType.newBuffer(haystackBytes);
+
+        kmpFactory = SearchProcessorFactory.newKmpSearchProcessorFactory(needleBytes);
+        ahoCorasicFactory = MultiSearchProcessorFactory.newAhoCorasicSearchProcessorFactory(needleBytes);
+
+        if (needleBytes.length <= 64) {
+            shiftingBitMaskFactory = SearchProcessorFactory.newShiftingBitMaskSearchProcessorFactory(needleBytes);
+        }
+    }
+
+    @TearDown
+    public void teardown() {
+        needle.release();
+        haystack.release();
+    }
+
+    @Benchmark
+    @CompilerControl(Mode.DONT_INLINE)
+    public int indexOf() {
+        return ByteBufUtil.indexOf(needle, haystack);
+    }
+
+    @Benchmark
+    @CompilerControl(Mode.DONT_INLINE)
+    public int kmp() {
+        return haystack.forEachByte(kmpFactory.newSearchProcessor());
```

Comment:
It cannot be made less GC hungry by caching the search processor? 
each benchmarked method could include the jump table initialization and needle clone too (thinking of kmp here)
Escape analysis probably won't even kick-in due to the size of needle array

Reply:
Sure, I can make the runs allocation free by reusing the search processor (resetting its state is trivial for all algos). Or is it something else that you had in mind?

Reply:
My new benchmark suite will reuse (reset) the `SearchProcessor` so each invocation will be allocation free.